flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java

```diff
@@ -0,0 +1,672 @@
+            boolean isParallelismConfigured =
+                    chainedStreamNodes.stream().anyMatch(StreamNode::isParallelismConfigured);
+            if (!isParallelismConfigured && streamGraph.isAutoParallelismEnabled()) {
+                chainedStreamNodes.forEach(
+                        n -> n.setParallelism(ExecutionConfig.PARALLELISM_DEFAULT, false));
+            }
+        }
+        // We calculate forward group by a set of chained stream nodes, and use the start node to
+        // identify the chain group.
+
+        // The value are start nod of upstream chain groups that connect target
+        // chain group by forward edge.
+        final Map<StreamNode, Set<StreamNode>> chainGroupToProducers = new LinkedHashMap<>();
+
+        for (StreamNode startNode : topologicallySortedChainedStreamNodesMap.keySet()) {
+            int startNodeId = startNode.getId();
+            // All downstream node connect to current chain group with forward partitioner.
+            Set<StreamNode> forwardConsumers =
+                    transitiveNonChainableOutEdgesMap.get(startNodeId).stream()
+                            .filter(
+                                    edge ->
+                                            edge.getPartitioner()
+                                                    .getClass()
+                                                    .equals(ForwardPartitioner.class))
+                            .map(StreamEdge::getTargetId)
+                            .map(streamGraph::getStreamNode)
+                            .collect(Collectors.toSet());
+            for (StreamNode forwardConsumer : forwardConsumers) {
+                chainGroupToProducers
+                        .computeIfAbsent(forwardConsumer, ignored -> new HashSet<>())
+                        .add(streamGraph.getStreamNode(startNodeId));
+            }
+        }
+
+        final Map<Integer, StreamNodeForwardGroup> forwardGroupsByStartNodeId =
+                ForwardGroupComputeUtil.computeStreamNodeForwardGroup(
+                        topologicallySortedChainedStreamNodesMap,
+                        startNode ->
+                                chainGroupToProducers.getOrDefault(
+                                        startNode, Collections.emptySet()));
+
+        forwardGroupsByStartNodeId.forEach(
+                (startNodeId, forwardGroup) -> {
+                    this.startAndEndNodeIdToForwardGroupMap.put(startNodeId, forwardGroup);
+                    // Gets end node of current chain group by transitiveNonChainableOutEdgesMap
+                    transitiveNonChainableOutEdgesMap
+                            .get(startNodeId)
+                            .forEach(
+                                    streamEdge -> {
+                                        this.startAndEndNodeIdToForwardGroupMap.put(
+                                                streamEdge.getSourceId(), forwardGroup);
```

Comment:
streamEdge -> this.startAndEndNodeIdToForwardGroupMap.put(
                                            streamEdge.getSourceId(), forwardGroup)